File: common-js/RequestHandler.js
const { Response } = require('./Message');
const MessageContextHolder = require('./MessageContextHolder');

class RequestHandler {
    constructor(config, client) {
        this.config = config;
        this.client = client;
        this.routes = new Map();
    }

    handle(uri, handlerFunc) {
        this.routes.set(uri, handlerFunc);
    }

    async dispatch(jsonPayload) {
        let msg;
        try {
            msg = JSON.parse(jsonPayload);
        } catch (e) {
            console.error("failed to parse kafka message", e);
            return;
        }

        if (this.config.timeDiffIgnoreMs > 0 && msg.timestamp) {
            const age = Date.now() - msg.timestamp;
            if (age > this.config.timeDiffIgnoreMs) {
                console.warn(`dropping stale message msgId: ${msg.messageId} age: ${age}`);
                return;
            }
        }

        const context = {
            transactionId: msg.transactionId,
            messageId: msg.messageId,
            sourceId: msg.sourceId,
            uri: msg.uri
        };

        // Run within Async context block to emulate ThreadLocal boundaries
        await MessageContextHolder.run(context, async () => {
            const start = Date.now();
            try {
                const handler = this.routes.get(msg.uri);
                if (!handler) {
                    throw new Error(`URI_NOT_FOUND: ${msg.uri}`);
                }

                let data = handler(msg);
                if (data instanceof Promise) {
                    data = await data;
                }
                
                if (data !== undefined && msg.responseDestination) {
                    const resp = Response.dataResponse(data);
                    await this.client.sendResponse(msg, resp);
                }

                console.log(`← response OK uri: ${msg.uri} took: ${Date.now() - start}ms`);

            } catch (err) {
                console.warn(`← response ERROR uri: ${msg.uri} took: ${Date.now() - start}ms error: ${err.message}`);
                
                if (msg.responseDestination) {
                    try {
                        const resp = Response.systemErrorResponse('GENERAL_ERROR');
                        await this.client.sendResponse(msg, resp);
                    } catch (sendErr) {
                        console.error("failed to send error response", sendErr);
                    }
                }
            }
        });
    }
}

module.exports = RequestHandler;

File: common-java/src/main/java/com/common/kafka/RequestHandler.java
package com.common.kafka;

import com.common.errors.ErrorHandlerUtils;
import com.common.message.Message;
import com.common.message.Response;
import com.common.utils.JsonUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public class RequestHandler {
    private static final Logger log = LoggerFactory.getLogger(RequestHandler.class);
    
    private final KafkaConfig config;
    private final KafkaClient client;
    private final Map<String, Function<Message, Object>> routes = new HashMap<>();

    public RequestHandler(KafkaConfig config, KafkaClient client) {
        this.config = config;
        this.client = client;
    }

    public void handle(String uri, Function<Message, Object> handler) {
        routes.put(uri, handler);
    }

    public void dispatch(String jsonPayload) {
        long start = System.currentTimeMillis();
        Message msg;
        try {
            msg = JsonUtils.getMapper().readValue(jsonPayload, Message.class);
        } catch (Exception e) {
            log.error("failed to parse kafka message", e);
            return;
        }

        if (config.getTimeDiffIgnoreMs() > 0 && msg.getTimestamp() != null) {
            long age = System.currentTimeMillis() - msg.getTimestamp();
            if (age > config.getTimeDiffIgnoreMs()) {
                log.warn("dropping stale message msgId: {} age: {}", msg.getMessageId(), age);
                return;
            }
        }

        try {
            MessageContextHolder.setTransactionId(msg.getTransactionId());
            MessageContextHolder.setMessageId(msg.getMessageId());
            MessageContextHolder.setSourceId(msg.getSourceId());
            MessageContextHolder.setUri(msg.getUri());

            Function<Message, Object> handler = routes.get(msg.getUri());
            if (handler == null) {
                throw com.common.errors.GeneralError.newUriNotFound(msg.getUri());
            }

            Object data = handler.apply(msg);
            
            if (data != null && msg.getResponseDestination() != null) {
                Response resp = Response.dataResponse(data);
                client.sendResponse(msg, resp);
            }
            
            log.info("← response OK uri: {} took: {}ms", msg.getUri(), (System.currentTimeMillis() - start));
        } catch (Throwable err) {
            log.warn("← response ERROR uri: {} took: {}ms error: {}", msg.getUri(), (System.currentTimeMillis() - start), err.getMessage());
            if (msg.getResponseDestination() != null) {
                try {
                    Response resp = ErrorHandlerUtils.toResponse(err);
                    client.sendResponse(msg, resp);
                } catch (Exception sendErr) {
                    log.error("failed to send error response", sendErr);
                }
            }
        } finally {
            MessageContextHolder.clear();
        }
    }
}
